theano/sandbox/cuda/opt.py

```diff
@@ -944,6 +944,35 @@ def local_gpu_reshape(node):
     return False
 
 
+@local_optimizer([GpuReshape])
+def local_gpu_reshape_chain(node):
+    """
+    GuReshape(GpuReshape(shape1),shape2) -> GpuReshape(shape2)
+
+    """
+    if not tensor.opt.opt.check_chain(node, GpuReshape, GpuReshape):
+        return False
+
+    # TODO: this can permit a failing program to run by eliminating
+    #       the lower reshape
+    rval = node.op(node.inputs[0].owner.inputs[0], node.inputs[1])
+    # It might happen that the desired output of this node has a broadcastable
+    # pattern that does not match that of 'rval'. This is when originally, we
+    # were able to figure out that one of the dimensions of the reshape is one,
+    # but some other transformation replaced the shape by one for which this
+    # cannot be guessed.
+    # We should try to figure out why we lost the information about this
+    # constant value... but in the meantime, better not apply this
+    # optimization.
+    if rval.broadcastable == node.outputs[0].broadcastable:
+        return [rval]
+    else:
+        return False
+gpu_cut_copies.register('cut_local_gpu_reshape_chain',
```

Comment:
This is a local optimizer. final opt are global optimizer. Also, this
should be in the equilibrium, no don't use any parameter.

On Thu, Sep 3, 2015 at 2:10 PM, Xavier Bouthillier <notifications@github.com
> wrote:

> In theano/sandbox/cuda/opt.py
> <https://github.com/Theano/Theano/pull/3357#discussion_r38677380>:
>
> > +    # TODO: this can permit a failing program to run by eliminating
> > +    #       the lower reshape
> > +    rval = node.op(node.inputs[0].owner.inputs[0], node.inputs[1])
> > +    # It might happen that the desired output of this node has a broadcastable
> > +    # pattern that does not match that of 'rval'. This is when originally, we
> > +    # were able to figure out that one of the dimensions of the reshape is one,
> > +    # but some other transformation replaced the shape by one for which this
> > +    # cannot be guessed.
> > +    # We should try to figure out why we lost the information about this
> > +    # constant value... but in the meantime, better not apply this
> > +    # optimization.
> > +    if rval.broadcastable == node.outputs[0].broadcastable:
> > +        return [rval]
> > +    else:
> > +        return False
> > +gpu_cut_copies.register('cut_local_gpu_reshape_chain',
>
> Should we use register_opt(final_opt=True)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3357/files#r38677380>.
>